src/main/kotlin/io/libp2p/core/security/secio/SecIoSecureChannel.kt

```diff
@@ -1,23 +1,108 @@
 import io.libp2p.core.protocol.ProtocolMatcher
-import io.libp2p.core.security.SecureChannel
+import io.libp2p.core.protocol.SecureChannel
+import io.libp2p.core.protocol.SecureChannelInitializerName
+import io.libp2p.core.types.toByteArray
+import io.libp2p.core.types.toByteBuf
+import io.netty.buffer.ByteBuf
+import io.netty.channel.ChannelHandlerContext
 import io.netty.channel.ChannelInboundHandlerAdapter
 import io.netty.channel.ChannelInitializer
-import io.netty.channel.socket.SocketChannel
+import io.netty.handler.codec.LengthFieldBasedFrameDecoder
+import io.netty.handler.codec.LengthFieldPrepender
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.async
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.channels.sendBlocking
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.withTimeout
+import java.util.concurrent.Executors
+import java.util.concurrent.atomic.AtomicInteger
+import io.netty.channel.Channel as NettyChannel
 
-class SecIoSecureChannel : SecureChannel {
-    override val matcher: ProtocolMatcher
-        get() = ProtocolMatcher(Mode.STRICT, name = "/secio/1.0.0")
 
-    override fun initializer(): ChannelInitializer<SocketChannel> =
-        object : ChannelInitializer<SocketChannel>() {
-            override fun initChannel(ch: SocketChannel) {
-                ch.pipeline().addLast(SecIoHandler())
+class SecIoSecureChannel(val localKey: PrivKey, val remotePeerId: PeerId? = null) :
+    SecureChannel {
+
+    private val HandshakeHandlerName = "SecIoHandshake"
+    private val HadshakeTimeout = 30 * 1000L
+
+    override val matcher = ProtocolMatcher(Mode.STRICT, name = "/secio/1.0.0")
+
+    override fun initializer(): ChannelInitializer<NettyChannel> =
+        object : ChannelInitializer<NettyChannel>() {
+            override fun initChannel(ch: NettyChannel) {
+                ch.pipeline().addBefore(SecureChannelInitializerName, "PacketLenEncoder", LengthFieldPrepender(4))
+                ch.pipeline().addBefore(SecureChannelInitializerName, "PacketLenDecoder",
+                    LengthFieldBasedFrameDecoder(Integer.MAX_VALUE, 0, 4, 0, 4))
+                ch.pipeline().addBefore(SecureChannelInitializerName, HandshakeHandlerName, SecIoHandshake())
             }
         }
 
-    class SecIoHandler: ChannelInboundHandlerAdapter()
+    inner class SecIoHandshake : ChannelInboundHandlerAdapter() {
+        private val kInChannel = Channel<ByteBuf>(1)
+        private var deferred: Deferred<Pair<SecioParams, SecioParams>>? = null
+        private val messageReadCount = AtomicInteger()
+        private var nonce: ByteArray? = null
+        private val executor = Executors.newSingleThreadExecutor()
+
+        override fun channelActive(ctx: ChannelHandlerContext) {
+            val negotiator = SecioHandshake(kInChannel, { buf -> writeAndFlush(ctx, buf) }, localKey, remotePeerId)
+
+            deferred = GlobalScope.async {
```

Comment:
I think there's a strong case to offer a coroutine-oriented façade to libp2p, but I would abstain from using coroutines inside the Netty domain because we're mixing two different threading models and discarding the threading/ordering guarantees that Netty nice provides for us. Personally I'd like to see three façades for the libp2p API:

* future-oriented (first one)
* coroutine-oriented (next one?)
* reactive streams (maybe, if coroutines haven't eaten the world).

Here I think we should be adding/removing handlers to/from the pipeline, and firing the appropriate events/reads/writes.

Reply:
Yep, I'm tending to agree with you regarding coroutines inside Netty handlers. Will replace this shortly ...
And totally agree on facades 👍